Introduce scheduler and base interfaces

yhabteab opened the pull request:
resolves #2

yhabteab committed:
Introduce trait `Timers`

yhabteab committed:
Introduce `Task` interface

yhabteab committed:
Introduce trait `Promises`

yhabteab committed:
Introduce trait `TaskProperties`

yhabteab committed:
Introduce `Scheduler` class

yhabteab committed:
Introduce interface `Frequency`

yhabteab committed:
Introduce class `Cron`

yhabteab committed:
Introduce file `composer.json`

lippserd reviewed (changes requested):
Please try to use `UuidInterface` instead of `TaskInterface` in `Promises` and `Timers`.

lippserd reviewed (changes requested):
In addition to the requested changes, I would like to see the methods adjusted that require access to a stored object, e.g. `unregisterPromise()`, `getTask()`, `detachTimer()`, etc. Some of them do nothing if the object does not exist, and some of them throw exceptions. I would always throw an `InvalidArgumentException` if the object does not exist. Otherwise, the caller has no guarantee what actually happened.

yhabteab committed:
TaskProperties: Raise an exception when required properties aren't set

lippserd committed:
Introduce events and remove logging

lippserd committed:
Code quality

- Do not pass by reference if it is not necessary
- PHPDoc and exception message improvements
- Use fewer variables and remove unused ones

lippserd committed:
Expect Task::run() to return ExtendedPromiseInterface

The scheduler requires to call always() on the returned promise.

lippserd committed:
Introduce events for task run and cancel

This also revises the traits so that they do not interfere with the loop
and do not cancel promises, as this is not their responsibility, but
that of concrete classes using the trait.

lippserd committed:
Document example usages of traits methods

lippserd committed:
Use static Loop functions

lippserd committed:
Document events
